File: my-module/android/src/main/java/expo/modules/mymodule/MyModule.kt
package expo.modules.mymodule

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.ui.ExpoUIView
import expo.modules.kotlin.records.recordFromMap
import expo.modules.ui.ModifierRegistry

class MyModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MyModule")

    Events("onChange")

    Constant("PI") {
      Math.PI
    }

    Function("hello") {
      "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { value: String ->
      sendEvent("onChange", mapOf(
        "value" to value
      ))
    }

    View(MyModuleView::class) {
      // Defines an event that the view can send to JavaScript.
      Events("onTap")
    }

    Class(MyModuleSharedObject::class) {
      Constructor {
        val instance = MyModuleSharedObject(appContext)
        return@Constructor instance
      }

      Property("count")
        .get { ref: MyModuleSharedObject ->
          ref.count
        }
        .set { ref: MyModuleSharedObject, count: Int ->
          ref.count = count
        }
    }

    ExpoUIView<MyModuleComposeViewProps>("MyModuleComposeView") {
      Content { props ->
        MyModuleComposeViewContent(props)
      }
    }

    OnCreate {
      ModifierRegistry.register("myModuleComposeModifier") { params, _, _, _ ->
        recordFromMap<MyModuleComposeModifierParams>(params).toModifier()
      }
    }
  }
}

File: my-module/ios/MyModule.swift
import ExpoModulesCore
import ExpoUI

public class MyModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MyModule")

    Events("onChange")

    Constant("PI") {
      Double.pi
    }

    Function("hello") {
      return "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { (value: String) in
      self.sendEvent("onChange", [
        "value": value
      ])
    }

    View(MyModuleView.self) {
      Events("onTap")
    }

    Class(MyModuleSharedObject.self) {
      Constructor { () -> MyModuleSharedObject in
        return MyModuleSharedObject()
      }

      Property("count") { (ref: MyModuleSharedObject) -> Int in
        return ref.count
      }
      .set { (ref: MyModuleSharedObject, count: Int) in
        ref.count = count
      }
    }

    ExpoUIView(MyModuleSwiftUIView.self)

    OnCreate {
      ViewModifierRegistry.register("myModuleSwiftUIModifier") { params, appContext, _ in
        return try MyModuleSwiftUIModifier(from: params, appContext: appContext)
      }
    }

    OnDestroy {
      ViewModifierRegistry.unregister("myModuleSwiftUIModifier")
    }
  }
}
